diff --git a/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF b/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF
index 1f1a40f0..a108e1b9 100644
--- a/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF
+++ b/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF
@@ -8,29 +8,29 @@ Bundle-RequiredExecutionEnvironment: JavaSE-17
 Bundle-Localization: plugin
 Import-Package: org.osgi.framework;version="1.3.0"
 Bundle-ActivationPolicy: lazy
-Require-Bundle: org.eclipse.core.runtime;bundle-version="3.12.0",
+Require-Bundle: org.eclipse.core.runtime;bundle-version="3.27.0",
  org.eclipse.core.resources,
  org.eclipse.core.net,
  org.eclipse.debug.core,
- org.eclipse.jdt.core,
+ org.eclipse.jdt.core;bundle-version="3.34.0",
  org.eclipse.ltk.core.refactoring,
- org.eclipse.text;bundle-version="3.6.0",
- org.eclipse.m2e.core;resolution:=optional,
- org.eclipse.m2e.jdt;resolution:=optional,
- org.eclipse.core.filebuffers;bundle-version="3.6.0",
- org.eclipse.m2e.maven.runtime;resolution:=optional,
- org.eclipse.buildship.core;bundle-version="1.0.18";resolution:=optional,
- org.eclipse.jdt.launching,
- org.eclipse.jdt.apt.core,
- org.eclipse.jdt.core.manipulation;bundle-version="1.8.0",
- com.google.gson;bundle-version="2.7.0",
+ org.eclipse.text;bundle-version="3.13.0",
+ org.eclipse.m2e.core;bundle-version="2.0.8";resolution:=optional,
+ org.eclipse.m2e.jdt;bundle-version="2.0.6";resolution:=optional,
+ org.eclipse.core.filebuffers;bundle-version="3.8.0",
+ org.eclipse.m2e.maven.runtime;bundle-version="3.9.100";resolution:=optional,
+ org.eclipse.buildship.core;bundle-version="3.1.7";resolution:=optional,
+ org.eclipse.jdt.launching;bundle-version="3.20.0",
+ org.eclipse.jdt.apt.core;bundle-version="3.8.0",
+ org.eclipse.jdt.core.manipulation;bundle-version="1.19.0",
+ com.google.gson;bundle-version="2.10.1",
  org.apache.commons.lang3;bundle-version="3.1.0",
  org.eclipse.lsp4j;bundle-version="[0.20.0.0,0.21.0)",
  org.eclipse.lsp4j.jsonrpc;bundle-version="[0.20.0,0.21.0)",
  org.eclipse.xtext.xbase.lib,
- org.eclipse.core.filesystem;bundle-version="1.7.0",
- org.eclipse.jdt.apt.pluggable.core;bundle-version="1.2.0";resolution:=optional,
- org.eclipse.m2e.apt.core;bundle-version="1.3.0";resolution:=optional
+ org.eclipse.core.filesystem;bundle-version="1.10.0",
+ org.eclipse.jdt.apt.pluggable.core;bundle-version="1.4.0";resolution:=optional,
+ org.eclipse.m2e.apt.core;bundle-version="2.2.0";resolution:=optional
 Export-Package: org.eclipse.jdt.ls.core.contentassist;x-friends:="org.eclipse.jdt.ls.tests",
  org.eclipse.jdt.ls.core.internal;x-friends:="org.eclipse.jdt.ls.tests,org.eclipse.jdt.ls.tests.syntaxserver",
  org.eclipse.jdt.ls.core.internal.cleanup;x-friends:="org.eclipse.jdt.ls.tests",
diff --git a/org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target b/org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target
index a8836e19..d53e01f2 100644
--- a/org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target
+++ b/org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target
@@ -29,16 +29,20 @@
             <unit id="org.eclipse.jdt.source.feature.group" version="0.0.0"/>
             <unit id="org.eclipse.sdk.feature.group" version="0.0.0"/>
             <unit id="org.mockito.mockito-core" version="0.0.0"/>
-            <repository location="https://download.eclipse.org/eclipse/updates/4.27/R-4.27-202303020300/"/>
+            <repository location="https://download.eclipse.org/eclipse/updates/4.28-I-builds"/>
         </location>
         <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
             <unit id="org.eclipse.xtext.xbase.lib" version="0.0.0"/>
-            <repository location="https://download.eclipse.org/releases/2022-12/"/>
+            <repository location="https://download.eclipse.org/releases/2023-03/"/>
         </location>
         <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
            <repository location="https://download.eclipse.org/lsp4j/updates/releases/0.20.0/"/>
            <unit id="org.eclipse.lsp4j" version="0.0.0"/>
         </location>
+	    <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
+		    <repository location="https://github.com/gayanper/org.gap.eclipse.jdt.eap/raw/gh-pages/p2/i-builds"/>
+		    <unit id="org.gap.eclipse.jdt.eap.patch.feature.feature.group" version="0.0.0"/>
+	    </location>
     </locations>
     <targetJRE path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17"/>
-</target>
+</target>
\ No newline at end of file
diff --git a/pom.xml b/pom.xml
index a3011c9e..8a704ddd 100644
--- a/pom.xml
+++ b/pom.xml
@@ -171,6 +171,11 @@
 								<ws>cocoa</ws>
 								<arch>x86_64</arch>
 							</environment>
+							<environment>
+								<os>macosx</os>
+								<ws>cocoa</ws>
+								<arch>aarch64</arch>
+							</environment>
 						</environments>
 						<dependency-resolution>
 							<!-- <optionalDependencies>ignore</optionalDependencies> -->
@@ -324,4 +329,4 @@
 			<url>https://repo.eclipse.org/content/repositories/jdtls-snapshots/</url>
 		</snapshotRepository>
 	</distributionManagement>
-</project>
+</project>
\ No newline at end of file
