diff --git a/java-extension/com.microsoft.java.test.plugin/META-INF/MANIFEST.MF b/java-extension/com.microsoft.java.test.plugin/META-INF/MANIFEST.MF
index 29694c5..488eacf 100644
--- a/java-extension/com.microsoft.java.test.plugin/META-INF/MANIFEST.MF
+++ b/java-extension/com.microsoft.java.test.plugin/META-INF/MANIFEST.MF
@@ -41,7 +41,7 @@ Require-Bundle: org.eclipse.jdt.core,
  org.apache.commons.lang3;bundle-version="3.1.0",
  com.google.gson;bundle-version="2.7.0",
  org.objectweb.asm;bundle-version="9.6.0",
- org.jacoco.core;bundle-version="0.8.11"
+ org.jacoco.core;bundle-version="0.8.12"
 Export-Package: com.microsoft.java.test.plugin.launchers;x-friends:="com.microsoft.java.test.plugin.test",
  com.microsoft.java.test.plugin.model;x-friends:="com.microsoft.java.test.plugin.test",
  com.microsoft.java.test.plugin.coverage;x-friends:="com.microsoft.java.test.plugin.test",
diff --git a/java-extension/com.microsoft.java.test.target/com.microsoft.java.test.tp.target b/java-extension/com.microsoft.java.test.target/com.microsoft.java.test.tp.target
index d2b6087..5ea5c62 100644
--- a/java-extension/com.microsoft.java.test.target/com.microsoft.java.test.tp.target
+++ b/java-extension/com.microsoft.java.test.target/com.microsoft.java.test.tp.target
@@ -48,7 +48,7 @@
                 <dependency>
                     <groupId>org.jacoco</groupId>
                     <artifactId>org.jacoco.core</artifactId>
-                    <version>0.8.11</version>
+                    <version>0.8.12</version>
                 </dependency>
             </dependencies>
         </location>
diff --git a/package.json b/package.json
index a90e43c..6aaef39 100644
--- a/package.json
+++ b/package.json
@@ -71,7 +71,7 @@
             "./server/org.eclipse.jdt.junit4.runtime_1.3.0.v20220609-1843.jar",
             "./server/org.eclipse.jdt.junit5.runtime_1.1.100.v20220907-0450.jar",
             "./server/org.opentest4j_1.2.0.jar",
-            "./server/org.jacoco.core_0.8.11.202310140853.jar",
+            "./server/org.jacoco.core_0.8.12.202403310830.jar",
             "./server/com.microsoft.java.test.plugin-0.41.0.jar"
         ],
         "viewsWelcome": [
