From 3d111d461579151846b57f02f7e957846a0a7460 Mon Sep 17 00:00:00 2001
From: Gayan Perera <gayanper@gmail.com>
Date: Sat, 27 Jul 2024 11:46:05 +0200
Subject: [PATCH] switch back to java 22 as requirement for runtime

Signed-off-by: Gayan Perera <gayanper@gmail.com>
---
 src/requirements.ts | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/requirements.ts b/src/requirements.ts
index 51f441d..39f589b 100644
--- a/src/requirements.ts
+++ b/src/requirements.ts
@@ -10,7 +10,7 @@ import { logger } from './log';
 import { checkJavaPreferences } from './settings';
 import { listJdks, sortJdksBySource, sortJdksByVersion } from './jdkUtils';
 
-const REQUIRED_JDK_VERSION = 23;
+const REQUIRED_JDK_VERSION = 22;
 /* eslint-disable @typescript-eslint/naming-convention */
 export interface RequirementsData {
     tooling_jre: string;
-- 
2.39.3 (Apple Git-146)

